File: easy/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {

  public boolean isValid(String s) {
    if (s.length() % 2 != 0) {
      return false;
    }

    Stack<Character> stack = new Stack<Character>();

    for (char c : s.toCharArray()) {
      if (c == ')' || c == ']' || c == '}') {
        char top = stack.empty() ? ' ' : stack.pop();
        if (
          c == ')' &&
          top != '(' ||
          c == ']' &&
          top != '[' ||
          c == '}' &&
          top != '{'
        ) {
          return false;
        }
      } else {
        stack.push(c);
      }
    }
    return stack.isEmpty();
  }
}
// @lc code=end

File: 75 must do questions/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        
        parentheses = {'(':')', '{':'}','[':']'}
        stack = []

        for char in s:
            if char in parentheses:
                stack.append(char)
            elif len(stack) == 0 or parentheses[stack.pop()] != char:
                return False
        
        return len(stack) == 0

        
        
# @lc code=end

